qf-faucet-server/src/main.rs

```diff
@@ -0,0 +1,288 @@
+
+    // Check if there are recent transfers
+    let count: i64 = conn.query_row(
+        "SELECT COUNT(*) FROM transfers 
+        WHERE address = ?1 AND timestamp > ?2",
+        [address, format!("{}", two_hours_ago).as_str()],
+        |row| row.get(0),
+    )?;
+    
+    // Delete all previous transfers from this address that are older than the timeout
+    conn.execute(
+        "DELETE FROM transfers WHERE address = ?1 AND timestamp < ?2",
+        [address, format!("{}", two_hours_ago).as_str()],
+    )?;
+
+    Ok(count == 0)
+}
+
+async fn transfer_tokens(body: DripRequest, config: ServerConfig) -> Result<impl warp::Reply, warp::Rejection> {
+    let can_transfer = can_transfer(body.address.as_str(), config.timeout)
+        .map_err(|e| {
+            println!("Error checking transfer: {:?}", e);
+            warp::reject::custom(StorageError)
+        })?;
+    
+    if !can_transfer {
+        return Ok(warp::reply::json(&"You have already received tokens in the last 2 hours"));
+    }
+
+    let account_bytes = hex::decode(&body.address)
+        .map_err(|e| {
+            println!("Error: {:?}", e);
+            warp::reject::custom(SomeError)
+        })?;
+
+    let account_array: [u8; 32] = account_bytes[..32].try_into()
+        .map_err(|e| {
+            println!("Error: {:?}", e);
+            warp::reject::custom(SomeError)
+        })?;
+    let rpc_endpoint = config.rpc_url;
+    let api = OnlineClient::<PolkadotConfig>::from_url(rpc_endpoint)
+        .await
+        .map_err(|e| {
+            println!("Error: {:?}", e);
+            warp::reject::custom(SomeError)
+        })?;
+
+    // Create a keypair from the provided address
+    let mnemonic = env::var("MNEMONIC").expect("MNEMONIC is not set");
+    let phrase = Mnemonic::from_str(mnemonic.as_str()).map_err(|e| {
+        println!("Error: {:?}", e);
+        warp::reject::custom(SomeError)
+    })?;
+
+    let from;
+    if config.debug {
+        println!("Transfer from Alice");
+        from = dev::alice();
+    } else {
+        from = sr25519::Keypair::from_phrase(&phrase, None).map_err(|e| {
+            println!("Error: {:?}", e);
+            warp::reject::custom(SomeError)
+        })?;
+    }
+    
+    let dest = MultiAddress::Id(AccountId32(account_array));
+
+
+    let transfer = polkadot::tx().balances().transfer_keep_alive(dest, FAUCET_AMOUNT.into());
+
+    let events = api
+        .tx()
+        .sign_and_submit_then_watch_default(&transfer, &from)
+        .await
+        .map_err(|e| {
+            println!("Error Submit transfer: {:?}", e);
+            warp::reject::custom(TransferError)
+        })?
+        .wait_for_finalized_success()
+        .await
+        .map_err(|e| {
+            println!("Error transfer not finalized: {:?}", e);
+            warp::reject::custom(TransferError)
+        })?;
+    
+    let _ = store_transfer(body.address.as_str(), 2, format!("{}", events.extrinsic_hash()).as_str())
```

Comment:
I think this could lead to faucet abuse if we store transfer info AFTER the transfer is made.  

Right now, the flow looks like this:  
`can_transfer -> do_transfer -> store_transfer`  

What if `do_transfer` takes several seconds to complete? A single address could send multiple parallel requests, and `can_transfer` would accept all of them, wasting test tokens on a single bad actor.  

I suggest this approach instead:  
`can_transfer -> store_transfer -> do_transfer`  

We could even update the row inside `can_transfer` atomically to prevent any concurrency issues.